utils: introduce trie_writer

michoecho opened the pull request:
This is the first part of a larger project meant to implement a trie-based
index format, (most likely the same or almost the same as Cassandra's BTI).

This commits introduces trie_writer, a class responsible for turning a
stream of (key, value) pairs (already sorted by key) into a stream of
serializable nodes, such that:

1. Each node lies entirely within one page (guaranteed).
2. Parents are located in the same page as their children (best-effort).
3. Padding is minimized (best-effort).

It does mostly what you would expect a "sorted keys -> trie" builder to do.
The hard part is calculating the sizes of nodes (which, in a well-packed on-disk
format, depend on the exact offsets of the node from its children) and grouping
them into pages.

This implementation mostly follows Cassandra's design of the same thing.
There are some differences, though. Notable ones:

1. The writer operates on chains of characters, rather than single characters.

   In Cassandra's implementation, the writer creates one node per character.
   A single long key can be translated to thousands of nodes.
   We create only one node per key. (Actually we split very long keys into
   a few nodes, but that's arbitrary and beside the point).

   For BTI's partition key index this doesn't matter.
   Since it only stores a minimal unique prefix of each key,
   and the trie is very balanced (due to token randomness),
   the average number of new characters added per key is very close to 1 anyway.
   (And the string-based logic might actually be a small pessimization, since
   manipulating a 1-byte string might be costlier than manipulating a single byte).

   But the row index might store arbitrarily long entries, and in that case the
   character-based logic might result in catastrophically bad performance.
   For reference: when writing a partition index, the total processing cost
   of a single node in the trie_writer is on the order of 800 instructions.
   Total processing cost of a single tiny partition during a `upgradesstables`
   operation is on the order of 10000 instructions. A small INSERT is on the
   order of 40000 instructions.

   So processing a single 1000-character clustering key in the trie_writer
   could cost as much as 20 INSERTs, which is scary. Even 100-character keys
   can be very expensive. With extremely long keys like that, the string-based
   logic is more than ~100x cheaper than character-based logic.
   (Note that only *new* characters happen here. If two index entries share a
   prefix, that prefix is only processed once. And the index is only populated
   with the minimal prefix needed to distinguish neighbours. So in practice,
   long chains might not happen often. But still, they are possible).

   I don't know if it makes sense to care about this case, but I figured the
   potential for problems is too big to ignore, so I switched to chain-based logic.

2. In the (assumed to be rare) case when a grouped subtree turns out to be bigger
   than a full page after revising the estimate, Cassandra splits it in a
   different way than us.

For testability, there is some separation between the logic responsible
for turning a stream of keys into a stream of nodes, and the logic
responsible for turning a stream of nodes into a stream of bytes.
This commit only includes the first part. It doesn't implement the target
on-disk format yet.

The serialization logic is passed to trie_writer via a template parameter.

There is only one test added in this commit, which attempts to be exhaustive,
by testing all possible datasets up to some size. The run time of the test
grows exponentially with the parameter size. I picked a set of parameters
which runs fast enough while still being expressive enough to cover all
the logic. (I checked the code coverage). But I also tested it with greater parameters
on my own machine (and with DEVELOPER_BUILD enabled, which adds extra sanitization).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12530/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12530/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost
:small_blue_diamond: boost/trie_writer_test
#### Failed Tests (1/2725):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12530/testReport/junit/boost.trie_writer_test.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 1 hr 3 min
- Builder: spider5.cloudius-systems.com

avikivity commented on utils/trie_writer.hh:
```diff
@@ -0,0 +1,1124 @@
+
+// Enables code which is useful for debugging during development,
+// but too expensive to be compiled into release builds (even if dynamically disabled).
+#define DEVELOPER_BUILD 0
```
Please prefix it with TRIE_, the name is too globally scoped. If possible, make it a constexpr bool (just a preference on my part, not a requirement).

michoecho replied:
Done.

avikivity commented on utils/trie_writer.hh:
These definitely need to be functions.

michoecho replied:
Why? I don't see a difference, except that functions will add some extra compile time.

michoecho replied:
@avikivity Ping. I'd like to understand what you want.

avikivity replied:
Macros are a separate mini-language with different scoping rules. For example, trie_logger will refer to what a function sees, while in a macro it will refer to whatever the caller of the macro sees. So I'd like to minimize its use.

I don't believe the savings in compile time merit using two languages.

michoecho replied:
Alright.

avikivity commented on utils/trie_writer.hh:
Put the whole thing in some namespace.

avikivity commented on utils/trie_writer.hh:
If it's a counter and not a TIMEstamp, don't call it a timestamp.

michoecho replied:
Renamed from `_discard_timestamp` to `_discard_counter`.

avikivity commented on utils/trie_writer.hh:
Without yet knowing how this is to be used, I'm guessing std::stack (or just std::vector with linear searches, even small_vector) will be better here.

michoecho replied:
Maybe they are faster in practice, but in theory they could cause quadratic behavior, since there I have no theoretical bound on the size of this thing. I'd rather leave it as a map. Note that it's only for testing anyway.

avikivity replied:
Right, for testing it doesn't matter.

avikivity commented on utils/trie_writer.hh:
How is destructor management supposed to happen? Can we require that T have a trivial destructor?

avikivity replied:
You assert it below. Nicer as a requires.

michoecho replied:
Moved from `static_assert` to `requires`.

avikivity commented on utils/trie_writer.hh:
I think this should move to sstables/trie or something. utils/ is for something that could have come out of std or boost or something. This is super specific to sstables.

michoecho replied:
Moved to `sstables/`

avikivity commented on utils/trie_writer.hh:
What's the motivation for making trie_writer a template?

Testability would be a good answer.

avikivity commented on utils/trie_writer.hh:
It would aid readability if the type we ptr\<T[]> or buf_ptr\<T>, to emphasize that it's not a pointer to one T.

michoecho replied:
In v2, I modified it so that `ptr<T>` and `ptr<T[]>` can both be used. (Even though in my opinion that's overkill).

avikivity commented on test/boost/trie_writer_test.cc:
```diff
@@ -0,0 +1,264 @@
+    }
+    testlog.debug("Outputs: {}", outputs);
+    SCYLLA_ASSERT(std::ranges::equal(inputs, outputs));
+    SCYLLA_ASSERT(std::ranges::equal(payloads, output_payloads));
```
Wouldn't BOOST_REQUIRE be better here?

SCYLLA_ASSERT might change its behavior one day.

michoecho replied:
It started as BOOST_REQUIRE, then I modified it to SCYLLA_ASSERT during debugging and I forgot to change it back. I'll change it back, since that's the convention.

That said, BOOST_REQUIRE is somewhat useless when compared to the ASSERT, because it doesn't leave coredumps and doesn't stop the debugger, it just prints an error and throws an exception. To actually debug a failed BOOST_REQUIRE, you want to turn it into an assert anyway.

avikivity replied:
I have the debugger catch the exception that BOOST_REQUIRE throws.

michoecho replied:
It works sometimes, (and it would work for this test too, true), but many tests throw legitimate exceptions, and then `catch throw` isn't really usable.

avikivity replied:
I have it catch the actual exception type (in those cases). I don't claim it's good, but that's what we have now.

michoecho replied:
Fixed.

avikivity commented:
Looks good.

michoecho commented:
v2:
Addresed Avi's comments:
- Turned `EXPENSIVE_ASSERT` and `EXPENSIVE_LOG` from macros into functions.
- Turned `DEVELOPER_BUILD` from a macro into a `constexpr bool`.
- Renamed `SANITIZE_BUMP_ALLOCATOR` to `TRIE_SANITIZE_BUMP_ALLOCATOR`.
- Wrapped the contents of `trie_writer.hh` into `namespace trie`.
- Renamed `discard_timestamp` to `discard_counter`.
- Moved the static assertions in `bump_allocator::alloc` to a `requires` clause.
- Changed uses of `SCYLLA_ASSERT` in the test to `BOOST_REQUIRE`.
- Clarified some code comments.
Addressed sanitizer's complaint about malloc/new mismatch (buffer allocated with `aligned_alloc` was being freed via `delete`).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12675/)
#### Build Details:
- Duration: 19 min
- Builder: i-049fd4652abe0d55d (m5ad.8xlarge)

michoecho commented:
v2.1: modified `bump_allocator` so that you can write both `ptr<T>` and `ptr<T[]>` to emphasize that something points to a single element or to an array respectively.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12691/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12691/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost
:small_blue_diamond: boost/trie_writer_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12691/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4393/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1324/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12415/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12691/)
#### Build Details:
- Duration: 5 hr 59 min
- Builder: spider6.cloudius-systems.com

michoecho commented:
@avikivity I addressed your comments. You can check what happened to the macros and to `ptr<T[]>` and see if you like it.

By the way, did you look at the logic too, or did you only review the general programming style? I'm guessing it's the latter.

@paszkow @denesb I think it would be good if at least one person reviewed all trie-related code to a point of full understanding. If you agree, please designate some reviewer.

denesb commented:
@lkshminarayanan please review.

michoecho commented:
I got a request to show the rest of my BTI implementation for context, so I pushed a draft to https://github.com/michoecho/scylla/commits/bti_draft/.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12969/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12969/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost
:small_blue_diamond: boost/trie_writer_test
#### Failed Tests (1/13641):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12969/testReport/junit/boost.view_schema_test.test_create_and_alter_mv_with_ttl.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 1 hr 50 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
> I got a request to show the rest of my BTI implementation for context, so I pushed a draft to https://github.com/michoecho/scylla/commits/bti_draft/.

Better to publish it as a draft pull request so comments have an anchor.

michoecho commented:
> > I got a request to show the rest of my BTI implementation for context, so I pushed a draft to https://github.com/michoecho/scylla/commits/bti_draft/.
> 
> Better to publish it as a draft pull request so comments have an anchor.

In that case, posted https://github.com/scylladb/scylladb/pull/21491.

lkshminarayanan commented on sstables/trie_writer.hh:
Using `std::span<const std::byte>` here requires the bytes to be linearized. Could we use `managed_bytes_view` instead?

lkshminarayanan commented on sstables/trie_writer.hh:
Remove `bump_allocator` parameter as it is unused in the method.

lkshminarayanan reviewed (approved):
I have reviewed the PR and the implementations looks good! Couple of minor suggestions below.

mykaul commented:
@michoecho - what is the status of this PR?
